### PR TITLE
fix: last `--jobserver-auth` wins 

### LINKED ISSUE #66: The last `--jobserver-auth` flag should be used
The first `--jobserver-auth` it finds is used, instead of the last like the Make manual asks:

> Be aware that the `MAKEFLAGS` variable may contain multiple instances of the `--jobserver-auth=` option. Only the last instance is relevant.

```console
$ echo 'fn main() {}' | MAKEFLAGS='--jobserver-auth= --jobserver-auth=3,4' rustc -
warning: failed to connect to jobserver from environment variable `MAKEFLAGS="--jobserver-auth= --jobserver-auth=3,4"`: cannot parse jobserver environment variable value: expected `fifo:PATH` or `R,W`, found ``
  |
  = note: the build environment is likely misconfigured
```

From https://github.com/rust-lang/rust/issues/120515.

### PR DESCRIPTION
**Pull request by weihanglo**
From the GNU make manual[^1]:

> Be aware that the `MAKEFLAGS` variable may contain multiple
> instances of the `--jobserver-auth=` option.
> Only the last instance is relevant.

Hence this commit makes it so.

With this commit, `--jobserver-auth` also takes precedence over
`--jobserver-fds`, even when `--jobserver-fds` is the last instance
of these flags. This is made intentionally since `--jobserver-fds`
was an undocumented internal-only flag before `--jobserver-auth`
made public, according to this announcement[^2].

[^1]: https://www.gnu.org/software/make/manual/make.html#Job-Slots
[^2]: https://git.savannah.gnu.org/cgit/make.git/tree/NEWS?h=4.2#n31

fixes #66

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -588,9 +584,45 @@ impl HelperState {
     }
 }
 
+fn find_jobserver_auth(var: &str) -> Option<(&str, usize)> {
```
[the remark this answers is not in the document]
Comment (weihanglo) [reply src/lib.rs:601]: Updated. Here is how it looks like in rustdoc (with `--document-private-items`).

image

File: src/lib.rs [hunk 2]
```diff
@@ -588,9 +584,59 @@ impl HelperState {
     }
 }
 
+/// Finds the position of `--jobserver-auth=` flag in a given environment variable.
```
Comment (ojeda) [inline src/lib.rs:587]: I guess the change is intended to be minimal, but I wonder if the function could simply return the final slice needed (i.e. the value of the flag found) rather than a tuple`(arg, pos)` to then compute a length outside and then the slice and so on.

**Review by ojeda — COMMENTED** [review]
Thanks for solving this!

File: src/lib.rs [hunk 3]
```diff
@@ -588,9 +584,59 @@ impl HelperState {
...
+///
+/// Precedence rules:
+///
+/// * The last instance wins [^1].
+/// * `--jobserver-fds=` as a fallback when no `--jobserver-auth=` is present [^2].
+///
+/// [^1]: See ["GNU male manual: Sharing Job Slots with GNU make"](https://www.gnu.org/software/make/manual/make.html#Job-Slots)
```
[the remark this answers is not in the document]
Comment (weihanglo) [reply src/lib.rs]: What a crazy typo…

File: src/lib.rs (hunk 2 above)
Comment (weihanglo) [reply src/lib.rs]: Nice idea! Updated. Let me know if you have any concern :)

File: src/lib.rs [hunk 4]
```diff
@@ -588,9 +584,59 @@ impl HelperState {
...
+/// _"Be aware that the `MAKEFLAGS` variable may contain multiple instances of
+/// the `--jobserver-auth=` option. Only the last instance is relevant."_
+///
+/// [^2]: Refer to [the release announcement](https://lists.gnu.org/archive/html/info-gnu/2016-05/msg00013.html)
```
[the remark this answers is not in the document]
Comment (weihanglo) [reply src/lib.rs]: This is much better!

File: src/lib.rs [hunk 5]
```diff
@@ -588,9 +584,59 @@ impl HelperState {
...
     let _y = x.clone();
     std::mem::drop(x.into_helper_thread(|_| {}).unwrap());
 }
+
+#[test]
+fn test_find_jobserver_auth() {
+    let cases = [
+        (
+            "--jobserver-auth= --jobserver-auth=",
+            ("--jobserver-auth=", 18),
```
[the remark this answers is not in the document]
Comment (weihanglo) [reply src/lib.rs]: More tests added. Let me know if there are extra tests you want to see.

File: src/lib.rs (hunk 2 above)
Comment (ojeda) [reply src/lib.rs]: Looks good, thanks! It is much simpler.

File: src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (ojeda) [reply src/lib.rs]: Thanks! That is exactly what I meant, yeah.

If you want to have more tests, some more ideas are: "failing" cases (`None`), single flag cases,  one with a `=`, cases with the real values GNU Make uses...

[the remark this answers is not in the document]
Comment (weihanglo) [reply src/lib.rs]: Thanks. More tests are added :)

**Review by petrochenkov — APPROVED** [review]
